### PR TITLE
Report Startup Crashes

### PR DESCRIPTION
**Pull request by romtsn**
## :scroll: Description

* Write Startup Crash marker file to disk when storing the envelope (if a crash happened within 2 seconds (configurable) after the SDK init)
* Read this marker file in SendCachedFireAndForgetIntegration and block the execution of `init` in case it exists for up to 5 seconds (configurable) 
* Continue asynchronously in background in case the blocking flush times out 

## :bulb: Motivation and Context

https://github.com/getsentry/team-mobile/issues/12

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:94]: This would also work for the outbox path, so Hybrid SDKs can benefit from that

File: sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MyApplication.java
Comment (adinauer) [inline sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MyApplication.java:27]: Did you mean to add a comment explaining how to trigger a startup crash here or should this code be removed?

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (adinauer) [reply sentry/src/main/java/io/sentry/SentryOptions.java:380]: Yeah seems generic enough to be used by more than Android.

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (adinauer) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:61]: Could we run into a case where people want to disable the cache but still keep the startup crash feature on?

**Review by adinauer — COMMENTED** [review]
Left some questions. Maybe @philipphofmann also wants to take a quick look since he implemented https://github.com/getsentry/sentry-cocoa/pull/2220

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryOptions.java:1890]: Use `ICurrentDateProvider` so you can unit test it.

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:231]: Can throw `SecurityException`

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:233]: Can throw Can throw `SecurityException`

Comment (marandaneto) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:233]: Can throw Can throw `SecurityException`

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryOptions.java:380]: Consider refactoring this, `SentryPerformanceProvider` and `SentryAndroid` already set:
`AppStartState.getInstance().setAppStartTime(appStartMillis, appStartTime);`
Which should be pretty much similar, maybe reusing would make sense, if possible.

File: sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MyApplication.java
Comment (romtsn) [reply sentry-samples/sentry-samples-android/src/main/java/io/sentry/samples/android/MyApplication.java:27]: yeah I was just testing, we could remove it

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:100]: So this is blocking the execution of this integration, but there's a list of integrations, I am wondering if the order here matters now, let's say this is the 1st integration, all the following integrations will have to wait, which is fine, my question is, does it make sense to install the most important integrations first before blocking this one? (such as Uncaught exception).

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:61]: I guess we just say we don't support this case, because we need the envelope to be stored in cache in the first place.. supporting it would mean we either block on unhandled exception or send the envelopes in a different process, which is a whole different story

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:233]: Deleting this file here would return false when doing the same thing for the other outbox integration (because the marker does not exist anymore), I guess you'd need to control the deletion afterward after this one and the other integration gets blocked and flushed.
Also, there's the challenge of the timeout, since it's 2 integrations but a single Future.

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:100]: Yeah we already install them in the SentryOptions ctor before everything else runs, right?  https://github.com/getsentry/sentry-java/blob/54cebc81e18ff2f50da062cac751681e7e4e83d4/sentry/src/main/java/io/sentry/SentryOptions.java#L1822-L1844

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:100]: Sounds good, I'd also make sure that `AnrIntegration` is after the blocking calls, we don't want to trigger ANRs if that's the intention.

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (philipphofmann) [inline sentry/src/main/java/io/sentry/SentryOptions.java:378]: `m`: We didn't add this to the options on Cocoa. Do you expect many users need to tweak those settings?

**Review by philipphofmann — COMMENTED** [review]
Concept-wise, this looks OK; I didn't take a too close look at the code thought.

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SentryOptions.java:378]: I don't, but I think we should definitely give them this option, 'cause startup times varies from app to app, and 2s might be too much or too little for some or the other. 

Did you expose the `startupCrashFlushTimeoutMillis` to the users though? That one I think we should definitely make configurable, because it's essentially blocking the app

File: sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SendCachedEnvelopeFireAndForgetIntegration.java:100]: yep, it's installed afterwards, all good here 👍

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SentryOptions.java:380]: Yeah, it makes sense. At first I thought I will also cover desktop/backend apps, but then I realized that `SendCachedFireAndForget` is only applied for Android.

afaiu, desktop and backend apps don't need this, because they are covered by the `ShutdownHookIntegration`, which will run a blocking flush before terminating, so we're good there as well. 

I will change it to the `AppStartState` things here 👍

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:233]: hmm, but the outbox integration uses a different path, so it should be good? We also check for file existence the line above, before calling `delete`, so it's also not a problem I guess.

I didn't get the single Future part, mind to explain?

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryOptions.java:380]: Makes sense.

File: sentry/src/main/java/io/sentry/cache/EnvelopeCache.java
Comment (marandaneto) [reply sentry/src/main/java/io/sentry/cache/EnvelopeCache.java:233]: If each folder has its own marker, that is fine but it's 2 IOs for each file (read/delete) for the same feature, wondering if we can improve this by having only one marker, but still, we'd wait for the execution of both integrations.
Ignore the `Future` part for now.

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: yea, but for how long, that's the question :) we chose reasonable defaults, but I think it should be configurable. It won't hurt I guess?

Anyway, I can also remove it, but I don't see it a problem if it's there.

Comment (philipphofmann) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: I try to avoid adding new options as much as I can. If you add it once, you have to support it. Adding it later is easy.

Comment (romtsn) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: I think I would at least keep the flushTimeout exposed since it controls for (up to) how long we are actually blocking the app of the users. Let's maybe ask what others think

**Comment by markushi** [comment]
Looks good to me! Is there a reason why this feature is Android-only? I guess it could be useful for Java as well.

File: sentry/src/main/java/io/sentry/SentryOptions.java
Comment (romtsn) [reply sentry/src/main/java/io/sentry/SentryOptions.java]: As discussed internally we'll add a kill-switch boolean flag for this feature

**Comment by romtsn** [comment]
this one is ready for review now, I still need at least one approval 😄 @philipphofmann @marandaneto @adinauer

**Comment by philipphofmann** [comment]
@kevinrenskers, made a good point in https://github.com/getsentry/sentry-cocoa/issues/2283#issuecomment-1278983023 on why you should have to use `enableStartupCrashDetection`. Unless this feature is broken, I don't think there is a good use case for disabling this feature.

**Review by adinauer — APPROVED** [review]
LGTM; maybe update PR title

**Comment by philipphofmann** [comment]
Thank you, @romtsn, for shipping this feature and being patient with my comments about the options.

**Comment by romtsn** [comment]
thanks for the thorough review and prior art 👍

**Pull request merged**
